Mathlib/Tactic/ApplyRules.lean

```diff
@@ -0,0 +1,69 @@
+/-
+Copyright (c) 2022 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import Std.Util.TermUnsafe
+import Mathlib.Tactic.Repeat
+
+/-!
+# The `apply_rules` tactic
+
+The `apply_rules` tactic calls `apply` (with a specified set of lemmas) and `assumption`
+repeatedly, until no more applications are possible.
+-/
+
+
+namespace Mathlib.Tactic
+open Lean Meta Elab Tactic Term Parser.Tactic
+
+/--
+`applyFirst l`, for `l : List (TacticM Expr)`,
+tries to apply one of the lemmas generated by the tactics in `l` to the first goal, and
+fail if none succeeds.
+
+Implementation note:
+We pass a list of `TacticM Expr` rather than just `Expr`,
+because when `(← getMainGoal).apply e` succeeds, it can modify `e` via unification,
+blocking subsequent applications which would succeed but unify differently.
+This implementation ensures that the expressions are freshly generated for each
+invocation of `apply`.
```

Comment:
I think you want to use `withoutModifyingState` for this, or call `saveState`  and `restore` directly in the function.

Reply:
I don't think `withoutModifyingState` is useful here: we want the main goal to change! Instead we need to "protect" the `Expr` objects from being modified, which (just as in mathlib3) we can achieve by simply regenerating them. The test checking for this is at https://github.com/leanprover-community/mathlib4/pull/412/files#diff-4c95e5fe787e9a1f63803c90004f842ee25f9ef0caee7d388581a3d17b05c030R14.

Reply:
If you use `saveState` and `restore` you can control whether to commit the changes to the metavariable context or not. The Expr objects themselves never change, but I assume you are talking about assignments to contained metavariables.

Reply:
Thanks. I didn't know we could separately track changes to the metavariable context. I'll try this approach.

Reply:
Curiously the test from mathlib3 that checked for correct behaviour when we repeatedly apply something with a metavariable seems to just work even with no special handling of the metavariable context (or regenerating the `Expr`s). This makes everything nice and simple.